router/src/gossip/mod.rs

```diff
@@ -52,3 +64,205 @@ pub mod traits;
 
 #[cfg(test)]
 mod mock_schema_broadcast;
+
+#[cfg(test)]
+mod tests {
+    use std::{collections::BTreeMap, sync::Arc, time::Duration};
+
+    use async_trait::async_trait;
+    use data_types::{
+        partition_template::{
+            test_table_partition_override, NamespacePartitionTemplateOverride,
+            TablePartitionTemplateOverride, PARTITION_BY_DAY_PROTO,
+        },
+        Column, ColumnId, ColumnsByName, NamespaceId, NamespaceName, NamespaceSchema, TableId,
+        TableSchema,
+    };
+    use gossip::Dispatcher;
+    use test_helpers::timeout::FutureTimeout;
+    use tokio::sync::Mutex;
+
+    use crate::namespace_cache::{MemoryNamespaceCache, NamespaceCache};
+
+    use super::{
+        dispatcher::GossipMessageDispatcher, namespace_cache::NamespaceSchemaGossip,
+        schema_change_observer::SchemaChangeObserver, traits::SchemaBroadcast,
+    };
+
+    #[derive(Debug, Default)]
+    struct GossipPipe {
+        dispatcher: Mutex<Option<GossipMessageDispatcher>>,
+    }
+
+    impl GossipPipe {
+        async fn set_dispatcher(&self, dispatcher: GossipMessageDispatcher) {
+            *self.dispatcher.lock().await = Some(dispatcher);
+        }
+    }
+
+    #[async_trait]
+    impl SchemaBroadcast for Arc<GossipPipe> {
+        async fn broadcast(&self, payload: Vec<u8>) {
+            self.dispatcher
+                .lock()
+                .await
+                .as_mut()
+                .unwrap()
+                .dispatch(payload.into())
+                .with_timeout_panic(Duration::from_secs(5))
+                .await;
+        }
+    }
+
+    // Place a new namespace with a table and column into node A, and check it
+    // becomes readable on node B.
+    //
+    // This is an integration test of the various schema gossip components.
+    #[tokio::test]
+    async fn test_integration() {
+        // Two adaptors that will plug one "node" into the other.
+        let gossip_a = Arc::new(GossipPipe::default());
+        let gossip_b = Arc::new(GossipPipe::default());
+
+        // Setup a cache for node A and wrap it in the gossip layer.
+        let node_a_cache = Arc::new(MemoryNamespaceCache::default());
+        let dispatcher_a = Arc::new(NamespaceSchemaGossip::new(Arc::clone(&node_a_cache)));
+        let dispatcher_a = GossipMessageDispatcher::new(dispatcher_a, 100);
+        let node_a = SchemaChangeObserver::new(Arc::clone(&node_a_cache), Arc::clone(&gossip_b));
+
+        // Setup a cache for node B.
+
+        let node_b_cache = Arc::new(MemoryNamespaceCache::default());
+        let dispatcher_b = Arc::new(NamespaceSchemaGossip::new(Arc::clone(&node_b_cache)));
+        let dispatcher_b = GossipMessageDispatcher::new(dispatcher_b, 100);
+        let node_b = SchemaChangeObserver::new(Arc::clone(&node_b_cache), Arc::clone(&gossip_b));
```

Comment:
`node_a` is gossiping to `gossip_b` (on line 131), and `node_b` is also gossiping to `gossip_b`? Is that right or should this line be:

```suggestion
        let node_b = SchemaChangeObserver::new(Arc::clone(&node_b_cache), Arc::clone(&gossip_a));
```

It doesn't really matter for this test since it's only asserting that a -> b, but I think b -> a would fail? Also the same thing is on line 213, maybe extract this setup into a function?